### PR TITLE
[AIRFLOW-1491] Recover celery queue on restart

### PR DESCRIPTION
**Pull request by bolkedebruin**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1491

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

If the scheduler (re)starts it will start from a clean slate.
This is not a problem for a synchronous executor (e.g. LocalExecutor),
but it is for an asynchronous one as worker might still be executing
tasks, while the scheduler will requeue them.

This patch makes sure the queue is recovered from the workers, by
inpsecting them.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Unit tests for celery added. Working on the recovery method testing

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

cc: @gwax @saguziel

### CODE REVIEW

**Comment by bolkedebruin** [comment]
Please note that this is not done yet. app.control.inspect() behaviour is... unpredictable. In other words I cannot get it to consistently  report queued tasks. It seems likely that celery is not aware of tasks that are in the queue (ie. the mq). It might need a different approach (storing UUID in a de table and Maybe query those, asking the broker directly etc)

@gwax @saguziel if you guys have any ideas that would be appreciated.

**Comment by saguziel** [comment]
Does recovering the queue actually get us much? I can see how it would be useful in the future, but it's probably more important to recover the `jobs` table which has LocalTaskJobs since this has the job heartbeat logic.

**Comment by saguziel** [comment]
I think it's pretty safe to double queue tasks as long as it isn't too frequent, but restarting the scheduler can cause running tasks that died to not get restarted

**Comment by bolkedebruin** [comment]
@gwax @saguziel This is a working implementation. The 'running' state of the executor is re-instated after a restart. I could also do the same for queued_tasks which would reinstate the whole state of the executor after a restart, effectively closing the gap that we have between executor and scheduler (tasks are set to QUEUED when moved to the executor, but there is a delay (ie. heartbeat) before it is send to the MQ thus a kill leaves a window in which a QUEUED task can get lost).

File: airflow/models.py [hunk 1]
```diff
@@ -2630,6 +2630,65 @@ def xcom_pull(
             include_prior_dates=include_prior_dates)
 
 
+class ExecutorQueue(Base):
+    __tablename__ = "executor_queue"
+
+    key = Column(String(ID_LEN), primary_key=True)
+    id = Column(String(ID_LEN), nullable=False)
```
Comment (saguziel) [inline airflow/models.py:2637]: I think that ExecutorQueue, like LocalTaskJobs, should include the hostname.

File: airflow/models.py [hunk 2]
```diff
@@ -2630,6 +2630,65 @@ def xcom_pull(
...
+    updated = Column(DateTime, nullable=False)
+
+
+class ExecutorQueueManager(dict):
+    def __init__(self, *args, **kwargs):
+        super(ExecutorQueueManager, self).__init__(*args, **kwargs)
+        self.session = settings.Session()
```
Comment (saguziel) [inline airflow/models.py:2644]: This pattern seems to be different than our current pattern. Any particular reason why?

File: airflow/models.py [hunk 3]
```diff
@@ -2630,6 +2630,65 @@ def xcom_pull(
...
+
+    def __getitem__(self, key):
+        qry = (self.session.query(ExecutorQueue)
```
Comment (saguziel) [inline airflow/models.py:2647]: Will these queries be expensive?

File: airflow/models.py [hunk 4]
```diff
@@ -2630,6 +2630,65 @@ def xcom_pull(
...
+    def __setitem__(self, key, value):
+        """
+        :param key: task_id
+        :param value: AsyncResult
+        :return:
+        """
+        task = ExecutorQueue()
+        task.key = key
+        task.id = value.id
+        task.updated = datetime.utcnow()
```
Comment (saguziel) [inline airflow/models.py:2662]: It might be better to use datetime.now() since everywhere else uses it ( maybe unfortunately).

**Comment by saguziel** [comment]
I think in `sync`, we should probably check if the task is  in state.reserved, and if it was enqueued reasonably long ago, we should requeue it.

File: airflow/models.py (hunk 3 above)
Comment (bolkedebruin) [reply airflow/models.py:2647]: No, key value lookup. No scans, nothing should be faster than this.

File: airflow/models.py (hunk 4 above)
Comment (bolkedebruin) [reply airflow/models.py:2662]: We should actually switch to utcnow() which is the canonical way to do it. We don't use "updated" anywhere at the moment so it suits future compatibility better (as we are saying to use UTC anyway).

**Comment by bolkedebruin** [comment]
On the `sync` part. I think we need to investigate this a little but further. It seems to be dependent on the broker (ie. redis vs rabbitmq) and visibility timeouts. It could also be fixed by doing the same for queued_tasks, maybe.

File: airflow/models.py (hunk 1 above)
Comment (bolkedebruin) [reply airflow/models.py:2637]: That doesn't make sense really, the queue is shared across all (potential) executors and across the celery cluster. Workers reporting back can happen at any of the executors and then the message is gone.

File: airflow/models.py (hunk 2 above)
Comment (bolkedebruin) [reply airflow/models.py:2644]: What part?

Comment (saguziel) [reply airflow/models.py:2644]: Session is usually the last arg to functions.

File: airflow/models.py (hunk 3 above)
Comment (saguziel) [reply airflow/models.py:2647]: Still has roundtrip time

File: airflow/models.py (hunk 4 above)
Comment (saguziel) [reply airflow/models.py:2662]: It's kind of inconsistent in that sense. It might make more sense to migrate everything at once.

**Comment by saguziel** [comment]
I don't think it's really related to visibility timeouts. FWIW, visibility timeouts don't cause incorrect behavior, just pollute logs. I think Redis does have the possibility of losing messages while Rabbit does not (which is why Redis has visibility timeout).

File: airflow/models.py (hunk 4 above)
Comment (bolkedebruin) [reply airflow/models.py:2662]: Agreed. But we are mixing anyway. But a move to utcnow everywhere makes sense. There was a pr out for that but we never accepted it. I might revive that one for 1.9. @criccomini if I do that in the next 2 days would you accept that?

Comment (criccomini) [reply airflow/models.py:2662]: @bolkedebruin Ya, if it's merged in by EOW, we can get it into 1.9

**Comment by SamWildmo** [comment]
Is this still an issue in 1.10? I'm considring switching  to celery and wondring if its smart due to this bug which hasnt been fixed yet?

**Pull request closed without merging**